File: backend/src/main/java/com/sigtm/backend/motocicleta/MotocicletaService.java
package com.sigtm.backend.motocicleta;

import com.sigtm.backend.cliente.Cliente;
import com.sigtm.backend.cliente.ClienteService;
import com.sigtm.backend.common.ForbiddenOperationException;
import com.sigtm.backend.common.ResourceNotFoundException;
import com.sigtm.backend.motocicleta.dto.MotocicletaRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

/**
 * Unica fuente de verdad para leer/escribir motos. El modulo "Portal del cliente"
 * NO duplica esta logica: reutiliza estos mismos metodos con el filtro de propiedad.
 */
@Service
public class MotocicletaService {

    private final MotocicletaRepository motocicletaRepository;
    private final ClienteService clienteService;

    public MotocicletaService(MotocicletaRepository motocicletaRepository, ClienteService clienteService) {
        this.motocicletaRepository = motocicletaRepository;
        this.clienteService = clienteService;
    }

    public List<Motocicleta> listarTodas() {
        return motocicletaRepository.findAll();
    }

    public List<Motocicleta> listarPorCliente(Long clienteId) {
        return motocicletaRepository.findByClienteId(clienteId);
    }

    public Motocicleta obtenerPorId(Long id) {
        return motocicletaRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Motocicleta no encontrada: " + id));
    }

    /** Usado por el portal del cliente: lanza 403 si la moto no es de este cliente (previene IDOR). */
    public Motocicleta obtenerVerificandoPropietario(Long motoId, Long clienteId) {
        Motocicleta moto = obtenerPorId(motoId);
        if (!moto.getCliente().getId().equals(clienteId)) {
            throw new ForbiddenOperationException("Esta moto no pertenece al cliente autenticado.");
        }
        return moto;
    }

    @Transactional
    public Motocicleta registrar(MotocicletaRequest request) {
        Cliente cliente = clienteService.obtenerPorId(request.clienteId());

        Motocicleta moto = new Motocicleta();
        moto.setCliente(cliente);
        aplicarDatos(moto, request);
        return motocicletaRepository.save(moto);
    }

    @Transactional
    public Motocicleta actualizar(Long id, MotocicletaRequest request) {
        Motocicleta moto = obtenerPorId(id);
        aplicarDatos(moto, request);
        return motocicletaRepository.save(moto);
    }

    private void aplicarDatos(Motocicleta moto, MotocicletaRequest request) {
        moto.setMarca(request.marca());
        moto.setModelo(request.modelo());
        moto.setPlaca(request.placa());
        moto.setAnio(request.anio());
        moto.setCilindraje(request.cilindraje());
        moto.setColor(request.color());
        moto.setChasis(request.chasis());
    }
}

File: src/services/motocicleta.service.js
const motocicletaRepository = require('../repositories/motocicleta.repository');
const clienteRepository = require('../repositories/cliente.repository');
const AppError = require('../utils/AppError');

async function crear({ clienteId, placa, marca, modelo, anio, color }) {
  const cliente = await clienteRepository.buscarPorId(clienteId);
  if (!cliente) {
    throw new AppError('El cliente asociado no existe.', 404);
  }

  const placaExistente = await motocicletaRepository.buscarPorPlaca(placa);
  if (placaExistente) {
    throw new AppError('Ya existe una motocicleta registrada con esa placa.', 409);
  }

  return motocicletaRepository.crear({ clienteId, placa, marca, modelo, anio, color });
}

async function listar() {
  return motocicletaRepository.listar();
}

async function obtenerPorId(id) {
  const moto = await motocicletaRepository.buscarPorId(id);
  if (!moto) {
    throw new AppError('Motocicleta no encontrada.', 404);
  }
  return moto;
}

async function listarPorCliente(clienteId) {
  const cliente = await clienteRepository.buscarPorId(clienteId);
  if (!cliente) {
    throw new AppError('El cliente no existe.', 404);
  }
  return motocicletaRepository.listarPorCliente(clienteId);
}

async function actualizar(id, datos) {
  await obtenerPorId(id);
  return motocicletaRepository.actualizar(id, datos);
}

async function eliminar(id) {
  await obtenerPorId(id);
  return motocicletaRepository.eliminar(id);
}

module.exports = {
  crear,
  listar,
  obtenerPorId,
  listarPorCliente,
  actualizar,
  eliminar,
};